### PR TITLE
Clean up dashboard URL management

### LINKED ISSUE #34447: Click behavior doesn't respect tab choice when redirecting to a dashboard
**Is your feature request related to a problem? Please describe.**
If you choose to use click behavior and set it up to go to a dashboard on a click, you can't select the tab, you'll always go to the first tab.

This is important if you want to do a click behavior on teh same dashboard, e.g. click on a card in the first tab and then go to the second tab.

EDIT: you can't go to a specific tab of a dashboard when doing click behavior from another dashboard :(

**Describe the solution you'd like**
Select the tab when you use click behavior

**Describe alternatives you've considered**
~You need to do a URL link to the tab and the category~ can't do that, as the URL will be hardcoded (/dashboard/...) and you're in a signed context where you need a signed token (base64 encoded)

**How important is this feature to you?**
Just figured out this by doing a click behavior

**Additional context**
NA

### LINKED ISSUE #43145: Fix url changes multiple times when switch Dashboard tab + PR follow-ups
**Context**

Convert DashboardControls to TS hooks #42777 causes a new issue. When a user switch Dashboard tab, url changes multiple times.

See https://github.com/metabase/metabase/pull/42777#issuecomment-2130092696 for more details.

- issue links: [Add static dashboards to embedding SDK](https://github.com/metabase/metabase/pull/41973)

**Dev notes**
From what I can see, the issue happens because we call history.replaceState from multiple places at the same time during Dashboard tab rendering:
* We update current active tab in https://github.com/metabase/metabase/blob/5e29f434a055a4d4413da16d7f1f7fe306c9440d/frontend/src/metabase/dashboard/components/DashboardTabs/use-sync-url-slug.ts#L81-L91

* and then we also set filters params to the URL as well in https://github.com/metabase/metabase/blob/5e29f434a055a4d4413da16d7f1f7fe306c9440d/frontend/src/metabase/hooks/use-synced-query-string.ts#L26-L32

The root cause seems to be that these calls override each other multiple times.

### LINKED ISSUE #44106: Click behavior to a second tab will generate additional history events on the browser
### Describe the bug

When you send the user to a second tab via a click behavior then the browser history contains an additional event so you can't use the browser back button to go back to the previous dashboard

### To Reproduce

1) create a question
2) create 2 dashboards, on one, create 2 tabs
3) add a question to the one that has only 1 tab and modify the click behaviour of the question to go to the dashboard with 2 tabs so it goes to the second tab
4) now click the link and it should take you to the second tab
5) now click on the back button, it will take you nowhere... when it should take you to the previous dashboard

### Expected behavior

It should take you to the dashboard that you used to go to the second dashboard, but it didn't, since it generated an additional event on the browser history

### Logs

NA

### Information about your Metabase installation

```JSON
v49.x
```

### Severity

P2

### Additional context

NA

### PR DESCRIPTION
**Pull request by kulyk**
Closes #34447, closes #44106, closes #43145

The PR reworks the way we manage dashboard URL query parameters to make it more predictable and reliable.
Not backporting as it seems quite risky, especially since it affects all kinds of dashboards (regular, public, embedded, x-rays)

### Problem overview

Overall, there are three key things we keep in the URL state (including hash) for dashboards:
- parameter values, example: `?category=Gadget&rating=4`
- current tab, example: `?tab=4-overview`
- sharable state: auto-refresh rate, theme, etc. Example: `#refresh=60`

We're syncing URL parameters and dashboard filter values inside the [`useSyncedQueryString` hook](https://github.com/metabase/metabase/blob/97cfe29e894d63733ce29e391d5ff8d5e98fe4d2/frontend/src/metabase/hooks/use-synced-query-string.ts#L6) used by the [`SyncedParametersList` component](https://github.com/metabase/metabase/blob/97cfe29e894d63733ce29e391d5ff8d5e98fe4d2/frontend/src/metabase/parameters/components/ParametersList/SyncedParametersList.tsx#L9). It's important to know that it uses `window.location` to read/update location state. OTOH, dashboard tab URL parameter is managed with `react-router` in the [`useSyncURLSlug` hook](https://github.com/metabase/metabase/blob/97cfe29e894d63733ce29e391d5ff8d5e98fe4d2/frontend/src/metabase/dashboard/components/DashboardTabs/use-sync-url-slug.ts#L54).

They're often trying to update the URL concurrently, and since they use different kinds of location states, they're usually out of sync and can overwrite one another in random order. That alone causes issues and makes it hard to reason about URL updates.

On top of that, when we open a dashboard with a `tab` query parameter (like `/dashboard/1?tab=2-overview`), there's a period of time when the [`getSelectedTabId` selector](https://github.com/metabase/metabase/blob/35c84771208aea7b67eabedf5d22786210f6dcc7/frontend/src/metabase/dashboard/selectors.ts#L487) returns `null` while dashboard is being loaded. That might confuse both hooks, and it's one of the reasons why #34447 is happening.

### Solution

1. `useSyncedQueryString` and `useSyncURLSlug` are removed in favor of a single `useDashboardUrlQuery` hook to avoid out-of-sync state and concurrent updates
2. More strict checks for when query parameters have to be updated (e.g., we won't try to update the URL if parameters order changes)
3. `getSelectedTabId` is updated to use the `tab` URL parameter from `window.location` while dashboard is being loaded

### CODE REVIEW

File: frontend/src/metabase/parameters/components/ParametersList/SyncedParametersList.tsx
Comment (kulyk) [inline frontend/src/metabase/parameters/components/ParametersList/SyncedParametersList.tsx:1]: From that point, `SyncedParametersList` is only meant to be used by the query builder. I've tried to get rid of it altogether and port that part of the QB to `react-router` as well, but `react-router-redux` actions trigger `initializeQB`, so it needs more work to get it right. I'm leaving this out of scope, it'd be worth to move `SyncedParamtersList` to `metabase/query_builder` though

File: frontend/src/metabase/public/containers/PublicOrEmbeddedDashboard/PublicOrEmbeddedDashboardPage.tsx [hunk 1]
```diff
@@ -13,7 +13,7 @@ import { PublicOrEmbeddedDashboard } from "./PublicOrEmbeddedDashboard";
 import { usePublicDashboardEndpoints } from "./WithPublicDashboardEndpoints";
 
 export const PublicOrEmbeddedDashboardPage = (props: WithRouterProps) => {
-  const { location } = props;
+  const { location, router } = props;
```
[the remark this answers is not in the document]
Comment (kulyk) [reply frontend/src/metabase/public/containers/PublicOrEmbeddedDashboard/PublicOrEmbeddedDashboardPage.tsx:16]: @deniskaber I couldn't find this component being used anywhere else except for `routes-public` and `routes-embed` (which are not part of the SDK, right?). OTOH, I can see another component called `PublicOrEmbeddedDashboard` being used in `embedding-sdk`, but it doesn't use neither `location` or `router` props

Is there another way `PublicOrEmbeddedDashboardPage` is used by the embedding SDK somehow?

**Review by alxnddr — APPROVED** [review]
Looks and works great 🏅

**Review by deniskaber — APPROVED** [review]
I checked most cases for embedding including the SDK.
Works good, thank you!

**Pull request merged**